kpops/pipeline.py

```diff
@@ -85,6 +98,67 @@ def to_yaml(self) -> str:
             self.model_dump(mode="json", by_alias=True, exclude_none=True)["components"]
         )
 
+    def build_execution_graph_from(
+        self,
+        components: list[PipelineComponent],
+        reverse: bool,
+        runner: Callable[[PipelineComponent], Coroutine],
+    ) -> Awaitable:
+        sub_graph_nodes = self.__get_graph_nodes(components)
+
+        async def run_parallel_tasks(tasks):
+            asyncio_tasks = []
+            for coroutine in tasks:
+                asyncio_tasks.append(asyncio.create_task(coroutine))
+            await asyncio.gather(*asyncio_tasks)
+
+        async def run_graph_tasks(pending_tasks: list[Awaitable]):
+            for pending_task in pending_tasks:
+                await pending_task
+
+        sub_graph = self.graph.subgraph(sub_graph_nodes)
+        transformed_graph = sub_graph.copy()
+
+        root_node = "root_node_bfs"
+        transformed_graph.add_node(root_node)
+
+        for node in sub_graph:
+            predecessors = list(sub_graph.predecessors(node))
+            if not predecessors:
+                transformed_graph.add_edge(root_node, node)
+
+        layers_graph = list(nx.bfs_layers(transformed_graph, root_node))
+
+        sorted_tasks = []
+        for layer in layers_graph[1:]:
+            parallel_tasks = self.__get_parallel_task_from(layer, runner)
+
+            if parallel_tasks:
+                sorted_tasks.append(run_parallel_tasks(parallel_tasks))
+
+        if reverse:
+            sorted_tasks.reverse()
+
+        return run_graph_tasks(sorted_tasks)
+
+    def __get_graph_nodes(self, components):
+        sub_graph_nodes = []
+        for component in components:
+            sub_graph_nodes.append(component.id)
+            sub_graph_nodes.extend(list(component.inputs))
+            sub_graph_nodes.extend(list(component.outputs))
+        return sub_graph_nodes
+
+    def __get_parallel_task_from(self, layer, runner):
+        parallel_tasks = []
+
+        for node_in_layer in layer:
+            component_node = self._component_index[node_in_layer]
+            if component_node.component is not None and not component_node.is_topic:
```

Comment:
is it possible that `is_topic=True` if `component is not None`? to me it look like the `is_topic` check is unnecessary.

Reply:
is not possible but I remember that I did it because the pyright wanted to xD

Reply:
what if you would remove the `is_topic` attribute entirely?